### PR TITLE
KAFKA-6383: complete shutdown for CREATED StreamThreads

### PR DESCRIPTION
**Pull request by rodesai**
When transitioning StreamThread from CREATED to PENDING_SHUTDOWN
free up resources from the caller, rather than the stream thread,
since in this case the stream thread was never actually started.

In KakfaStreams.close, shut down the streams threads from the
close thread. StreamThread.shutdown may now block, so call this
from the close thread so that the timeout is honored.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by rodesai** [comment]
@mjsax @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 1]
```diff
@@ -714,7 +720,10 @@ public StreamThread(final Time time,
     @Override
     public void run() {
         log.info("Starting");
-        setState(State.RUNNING);
+        if (!setState(State.RUNNING)) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Could this ever happen? I.e. instead of checking the lower-level tread state here, we already check the higher-level instance state that in `KafkaStreams#start` if the instance state is not `CREATED` we will fail and throw?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 2]
```diff
@@ -796,23 +796,23 @@ public synchronized boolean close(final long timeout, final TimeUnit timeUnit) {
...
-                globalStreamThread.shutdown();
-            }
-
             // wait for all threads to join in a separate thread;
             // save the current thread so that if it is a stream thread
             // we don't attempt to join it and cause a deadlock
             final Thread shutdownThread = new Thread(new Runnable() {
                 @Override
                 public void run() {
+                    // notify all the threads to stop; avoid deadlocks by stopping any
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Is it necessary to move this logic within the close thread? Maybe I'm still missing something..

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 3]
```diff
@@ -209,6 +211,10 @@ boolean setState(final State newState) {
             stateListener.onChange(this, state, oldState);
         }
 
+        if (oldState == State.CREATED && newState == State.PENDING_SHUTDOWN) {
+            completeShutdown(true);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I'm wondering if we can avoid injecting the thread logic in the `setState` function. One idea is to not allow the transition from `CREATED` to `PENDING_SHUTDOWN`, and hence the `setState(PENDING_SHUTDOWN)` in `KafkaStreams#shutdown()` will return false (note all other states except `DEAD` can transit to `PENDING_SHUTDOWN`), and then we can depend on the return value i.e. in `shutdown()`:

```
if (!setState(PENDING_SHUTDOWN) && state() == CREATED)) {
    completeShutdown(true);
}
```

**Review by guozhangwang — COMMENTED** [review]
@rodesai thanks for the patch, left some comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (rodesai) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: If we did that we could still race with with the StreamThread itself and do stuff in run() while the caller of shutdown() is in completeShutdown(). I do agree that putting this logic in setState is ugly. Another option would be to have a casState (compare and swap state) function so the caller can be sure which state transition its making.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (rodesai) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: If the client does:
```
KafkaStreams ks = ...
ks.start();
ks.close();
```
Isn't it possible that the StreamThread.shutdown() gets called, sees that the thread is still in CREATED state and starts shutting the stream thread down before the thread itself gets scheduled or gets far enough into run() ?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 2 above)
Comment (rodesai) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: My thinking was that since ThreadState.shutdown() can now call ThreadState.completeShutdown(), its now possible that ThreadState.shutdown() will block on the producer exiting (or something else exiting... it seems to be closing a lot of stuff). If this wait is indeterminately long, then we need to call shutdown() from the close thread to make sure we honor the timeout passed in to close().

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Ack, that makes sense.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Ack, makes sense.

**Comment by guozhangwang** [comment]
I think your suggestion of `casState` is more promising. More specifically, we can actually modify the `setState()` return type as the old state, and if the setState fails return `null`. Here is my reasoning:

There are two calling sequences that matters.

1) `StreamThread#run()` is never called since `KafkaStreams#start()` is not called. In this case we can let the caller thread of `KafkaStreams#shutdown()` to help trigger starting the stream threads after setting their state to `PENDING_SHUTDOWN`, we know that no other callers would call `thread#start()` again since it is guarded by the instance state.

2) `StreamThread#run()` is eventually called but the thread state is no longer `CREATED`.  In this case the thread should just proceed to call `completeShutdown`.

So it is still doable to only let the stream threads themselves to do the cleanup, to keep logic simpler to reason. To sum:

1) in `KafkaStreams#close()` we call `setState(PENDING_SHUTDOWN)` and then if it returns `CREATED` go ahead to call `thread.start()` after `thread.shutdown()`; if it returns others just call `thread.start()` (as is today); if it returns null it means the old state is `NOT_RUNNING` and it should not happen and hence we throw and fail.

2) in `KafkaStreams#start()` we call `setState(RUNNING)`, and if it returns null it means either `PENDING_SHUTDOWN` or `NOT_RUNNING`, and hence we do not need to proceed further, and just return.

3) in `StreamThread#run()` function if `setState(RUNNING)` returns null we can just proceed to `completeShutdown()`.

WDYT?

**Comment by rodesai** [comment]
That works. We can move the calls to shutdown back out of the clean up thread too.

That said, do you think it would be cleaner to move the
```
if (setState(PENDING_SHUTDOWN) == CREATED) {
    thread.start();
}
```
logic into StreamThread.shutdown(). This gives StreamThread cleaner semantics. As long as shutdown() gets called on it, everything gets cleaned up. The client doesn't have to worry about calling start() in that case.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: But couldn't we have a lock on the state within `completeShutdown()` ? Thus, if we are not running yet, and enter `completeShutdown()` we also don't allow to transit into RUNNING -- and if we did leave CREATED, we "cancel" `completeShutdown()` ?

**Review by mjsax — COMMENTED** [review]
We should add a test that calls `close()` without previous `start()` and pass in a mocked client to verify that those get closed (`KafkaStreams` allows to pass in a client supplier in it's constructor).

**Comment by guozhangwang** [comment]
> logic into StreamThread.shutdown(). This gives StreamThread cleaner semantics. As long as shutdown() gets called on it, everything gets cleaned up. The client doesn't have to worry about calling start() in that case.

Sounds good to me.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 4]
```diff
@@ -174,22 +174,24 @@ public State state() {
      * Sets the state
      * @param newState New state
      */
-    boolean setState(final State newState) {
-        final State oldState = state;
+    State setState(final State newState) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Nit: can we update JavaDoc with `@return`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 5]
```diff
@@ -274,7 +274,11 @@ void shutdown(final boolean clean) {
         standby.close(clean);
 
         // remove the changelog partitions from restore consumer
-        restoreConsumer.unsubscribe();
+        try {
+            restoreConsumer.unsubscribe();
+        } catch (final RuntimeException fatalException) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:279]: Why this change? AFAIK, `unsubscribe()` will never throw.

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 6]
```diff
@@ -113,6 +115,28 @@ public void testStateCloseAfterCreate() {
         Assert.assertEquals(KafkaStreams.State.NOT_RUNNING, streams.state());
     }
 
+    @Test
+    public void testStreamsCloseWithoutStart() throws Exception {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java]: nit: rename `shouldCleanupResourcesOnCloseWithoutPreviousStart()`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 7]
```diff
@@ -455,6 +455,35 @@ public void shouldShutdownTaskManagerOnClose() throws InterruptedException {
...
 
+    @Test
+    public void shouldShutdownTaskManagerOnCloseWithoutStart() {
+        final Consumer<byte[], byte[]> consumer = EasyMock.createNiceMock(Consumer.class);
+        final TaskManager taskManager = EasyMock.createNiceMock(TaskManager.class);
+        taskManager.shutdown(true);
+        EasyMock.expectLastCall();
+        EasyMock.replay(taskManager, consumer);
+
+        StreamThread.StreamsMetricsThreadImpl streamsMetrics = new StreamThread.StreamsMetricsThreadImpl(metrics, "", "", Collections.<String, String>emptyMap());
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: nit: add `final` and line too long

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 8]
```diff
@@ -455,6 +455,35 @@ public void shouldShutdownTaskManagerOnClose() throws InterruptedException {
...
+        final StreamThread thread = new StreamThread(mockTime,
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: nit: fix indention -- either move `mockTime` one line down or indent other parameters to align with `mockTime`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java [hunk 9]
```diff
@@ -455,6 +455,35 @@ public void shouldShutdownTaskManagerOnClose() throws InterruptedException {
...
+                null,
+                taskManager,
+                streamsMetrics,
+                internalTopologyBuilder,
+                clientId,
+                new LogContext(""));
+        thread.shutdown();
+        try {
+            thread.join(1000);
+        } catch (InterruptedException e) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 10]
```diff
@@ -1088,7 +1094,11 @@ private long computeLatency() {
      */
     public void shutdown() {
         log.info("Informed to shut down");
-        setState(State.PENDING_SHUTDOWN);
+        State oldState = setState(State.PENDING_SHUTDOWN);
+        if (oldState == State.CREATED) {
+            // Start so that we shutdown on the thread
+            this.start();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1101]: Why don't we call `completeShutdown(true)` directly instead via the redirect to `run()` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java (hunk 5 above)
Comment (rodesai) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:279]: It throws in the test since the consumer was never opened.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 10 above)
Comment (rodesai) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1101]: It seemed simpler to have all the shutdown stuff handled in the streams thread. I don't have a strong preference either way though.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1101]: This lgtm.

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk.
